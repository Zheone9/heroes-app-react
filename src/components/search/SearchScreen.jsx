import React, { useMemo } from "react";

import queryString from "query-string";

import { useLocation, useNavigate } from "react-router-dom";

import useForm from "../../hooks/useForm";
import HeroeCard from "../heroes/HeroeCard";
import { getHeroesByName } from "../../selectors/getHeroesByName";
import "../styles.css";

import useScrollPosition from "../../hooks/useScrollPos";
const SearchScreen = () => {
  const moreThanHalf = useScrollPosition("search");

  const location = useLocation();
  const navigate = useNavigate();

  const { q = "" } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({
    searchTerm: q,
  });

  const { searchTerm } = formValues;
  const handleClick = () => {
    window.scrollTo({ top: 0, behavior: "instant" });
  };
  const handleSearch = (e) => {
    e.preventDefault();
    if (searchTerm.trim().length === 0) return;

    navigate(`?q=${searchTerm}`);
  };
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  return (
    <div>
      <h1>SearchScreen</h1>
      <hr />
      <div className="row">
        <div>
          <h4>Search form</h4>

          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Search"
              className="form-control"
              name="searchTerm"
              onChange={handleInputChange}
              autoComplete="off"
              value={searchTerm}
            />
            <button
              type="submit"
              className="btn w-100 btn-outline-primary mt-1"
            >
              Search
            </button>
          </form>
        </div>
        <div>
          <h4>Results</h4>
          <hr />
          <div className="row row-cols-2 row-cols-md-4 g-4">
            {heroesFiltered.length === 0 && (
              <div
                className="alert alert-danger w-100 text-center"
                role="alert"
              >
                There is no a hero by the name of: {searchTerm}
              </div>
            )}

            {heroesFiltered.map((heroe) => (
              <HeroeCard key={heroe.id} {...heroe} />
            ))}
          </div>

          {moreThanHalf && (
            <i
              className="bi bi-arrow-up-circle-fill bsubir  animate__animated animate__fadeIn animate__faster"
              onClick={handleClick}
            ></i>
          )}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
