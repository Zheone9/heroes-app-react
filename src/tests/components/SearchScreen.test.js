import React from "react";

import { mount } from "enzyme";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import SearchScreen from "../../components/search/SearchScreen";
window.scrollTo = jest.fn();
const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("Pruebas en <SearchScreen/>", () => {
  beforeEach(() => {
    mockedUsedNavigate.mockReset();
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  test("debe de mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Routes>
          <Route path="search" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrar a batman y el input con el query string", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Routes>
          <Route path="search" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper.find(".alert").exists()).toBe(false);

    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrar un error si no se encuentra el heroe", () => {
    const heroe = "batmans23";

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${heroe}`]}>
        <Routes>
          <Route path="search" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      `There is no a hero by the name of: ${heroe}`
    );
  });

  test("debe de llamar el navigate al tocar el boton search", () => {
    const heroe = "batman";

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${heroe}`]}>
        <Routes>
          <Route path="search" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );
    wrapper.find("form").simulate("submit", {
      preventDefault() {},
    });
    wrapper
      .find("input")
      .simulate("change", { target: { name: "searchTerm", value: heroe } });

    expect(mockedUsedNavigate).toHaveBeenCalledWith(`?q=${heroe}`);
    expect(wrapper.find("input").prop("value")).toBe(heroe);
  });
});
