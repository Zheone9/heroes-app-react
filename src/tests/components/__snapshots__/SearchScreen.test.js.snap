// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> debe de mostrar a batman y el input con el query string 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div>
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchTerm"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value="batman"
        />
        <button
          className="btn w-100 btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div>
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="row row-cols-2 row-cols-md-4 g-4"
      >
        <div
          className="col"
        >
          <div
            className="card"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                <b>
                   
                  Batman
                </b>
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/heroe/dc-batman"
                onClick={[Function]}
              >
                 Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div>
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchTerm"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
        <button
          className="btn w-100 btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div>
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="row row-cols-2 row-cols-md-4 g-4"
      >
        <div
          className="alert alert-danger w-100 text-center"
          role="alert"
        >
          There is no a hero by the name of: 
        </div>
      </div>
    </div>
  </div>
</div>
`;
